### PR TITLE
Integration

### PR DESCRIPTION
**Pull request by richie1988**
1. ## Integration specs

## User index page:

- I can see the username of all other users.
- I can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.

## user show page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.

## User post index page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.

## Post show page:

- I can see the post's title.
- I can see who wrote the post.
- I can see how many comments it has.
- I can see how many likes it has.
- I can see the post body.
- I can see the username of each commentor.
- I can see the comment each commentor left.

## Before
ShareX_MAFl0MSYfo

## After
ShareX_m7S1aqB0i7

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -7,8 +7,19 @@ def index
...
+      # Handle the case when the post is not found, e.g., redirect to an error page
+      redirect_to user_posts_path(@user), alert: 'Post not found.'
+      return
+    end
+
     @comments = @post.comments.includes(:user).paginate(page: params[:page], per_page: 5)
+
+    respond_to do |format|
+      format.html { render template: 'posts/show' }
+    end
```
Comment (omar-muhamad) [inline app/controllers/posts_controller.rb]: - [ ] To solve the N+1 problem when fetching all posts and their comments for a user. The post should be like that `@posts = @user.posts.includes(:comments)`, not like you did above. Please fix that.

- [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=In%20your%20PR%20description%20include%20the%20screenshots%20of%20your%20app%27s%20logs%20before%20and%20after%20the%20fix), Please include the screenshots of your app's logs before and after the N+1 fix.

File: spec/views/users/index.html.erb_spec.rb [hunk 2]
```diff
@@ -1,9 +1,51 @@
...
+      @users.each do |u|
+        expect(page).to have_content('Number of Posts:')
+        expect(page).to have_content u.posts_counter
+      end
+    end
+
+    it 'When I click on a user, I am redirected to that user\'s show page.' do
+      first('a', text: @users[0].name).click
+      expect(page).to have_content @users[0].bio
+    end
```
Comment (omar-muhamad) [inline spec/views/users/index.html.erb_spec.rb]: - [ ] This is not the correct way for this test case. For this case, you should test the current path (user_path) when the user is clicked. So please fix that.

```suggestion
    it 'When I click on a user, I am redirected to that user\'s show page.' do
      click_link @user[0].name
      expect(page).to have_current_path user_path(@users[0])
    end
```

File: spec/views/users/show.html.erb_spec.rb [hunk 3]
```diff
@@ -1,9 +1,53 @@
...
+    it 'I can see the user"s bio' do
+      expect(page).to have_content 'My bio #landry !'
+    end
+
+    it 'I can see the user\'s first 3 posts.' do
+      expect(page).to have_content('This is my 0 post', wait: 10)
+      expect(page).to have_content('This is my 1 post', wait: 10)
+      expect(page).to have_content('This is my 2 post', wait: 10)
+    end
+
```
Comment (omar-muhamad) [inline spec/views/users/show.html.erb_spec.rb:41]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20a%20button%20that%20lets%20me%20view%20all%20of%20a%20user%27s%20posts.), Please add a test that asserts you can see a button that lets me view all of a user's posts.
- [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20a%20user%27s%20post%2C%20it%20redirects%20me%20to%20that%20post%27s%20show%20page.), Please add a test that redirects to the post's show page when you click on a user's post.

File: spec/views/users/show.html.erb_spec.rb [hunk 4]
```diff
@@ -1,9 +1,53 @@
...
+      visit user_path(@user.id)
+      # Use `find` instead of `click_on` to make it more robust
+      find('a', text: /See All Posts/i).click
+
+      # Check the content on the redirected page
+      (0..2).each do |i|
+        # Removed the `wait` option from `have_content`
+        expect(page).to have_content "This is my #{i} post"
+      end
+    end
```
Comment (omar-muhamad) [inline spec/views/users/show.html.erb_spec.rb]: - [ ] Same here, This is not the correct way to test this case. You should test the current path (user_posts_path) when the `see all posts` button is clicked, not testing a specific page content when the `see all posts` button is clicked. Please fix that.

File: spec/views/posts/index.html.erb_spec.rb [hunk 5]
```diff
@@ -1,9 +1,54 @@
...
+      Comment.create(post: @post, user: @user, text: "Comment #{i}")
+    end
+  end
+
+  after(:each) do
+    Comment.destroy_all
+    Post.destroy_all
+    User.destroy_all
+  end
+
```
Comment (omar-muhamad) [inline spec/views/posts/index.html.erb_spec.rb:18]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=post%20index%20page%3A-,I%20can%20see%20the%20user%27s%20profile%20picture.,-I%20can%20see), Please add a test that asserts you can see the user's profile picture.

File: spec/views/posts/index.html.erb_spec.rb [hunk 6]
```diff
@@ -1,9 +1,54 @@
...
+    visit user_posts_path(@user)
+    expect(page).to have_content @user.name
+  end
+
+  it 'I can see the number of posts the user has written.' do
+    visit user_posts_path(@user)
+    expect(page).to have_content 'Number of Posts: 1'
+  end
+
+  it 'I can see some of the post\'s body.' do
```
Comment (omar-muhamad) [inline spec/views/posts/index.html.erb_spec.rb:29]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20a%20post%27s%20title.), Please add a test that asserts you can see a post's title.
- [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20the%20first%20comments%20on%20a%20post.), Please add a test that asserts you can see the first comments on a post.

File: spec/views/posts/index.html.erb_spec.rb [hunk 7]
```diff
@@ -1,9 +1,54 @@
...
+  it 'I can see a section for pagination if there are more posts than fit on the view.' do
+    10.times do |i|
+      Post.create(author: @user, title: "Post #{i + 1}", text: "This is post number #{i + 1}")
+    end
+
+    visit user_posts_path(@user)
+
+    expect(page).to have_css('.pagination')
+    expect(page).to have_link('2')
   end
```
Comment (omar-muhamad) [inline spec/views/posts/index.html.erb_spec.rb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20post%2C%20it%20redirects%20me%20to%20that%20post%27s%20show%20page.), Please add a test that redirects to the post's show page when you click on a post.

File: spec/views/users/index.html.erb_spec.rb [hunk 8]
```diff
@@ -1,9 +1,51 @@
...
+        end
+      end
+    end
+
+    it 'I can see the number of posts each user has written.' do
+      @users.each do |u|
+        expect(page).to have_content('Number of Posts:')
+        expect(page).to have_content u.posts_counter
+      end
+    end
```
Comment (omar-muhamad) [inline spec/views/users/index.html.erb_spec.rb]: - [ ] There is one test case is failing, Please fix that and ensure that all tests are passing.

image

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @richie1988 and @martinkarugaba,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/views/users/show.html.erb_spec.rb (hunk 4 above)
Comment (richie1988) [reply spec/views/users/show.html.erb_spec.rb]: This part you are wrong its not the path its a content  cause it does not lead any were

File: spec/views/users/index.html.erb_spec.rb (hunk 8 above)
Comment (LionRouge1) [reply spec/views/users/index.html.erb_spec.rb]: - [ ] As mentioned by the previous reviewer the test to see the number of posts each user has written still failing. Kindly check it and make sure all the test are passing.
Screenshot from 2023-12-19 21-31-20

**Review by LionRouge1 — CHANGES_REQUESTED** [review]
Hi **Team**, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost

## Suggested changes :recycle: 
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

[OPTIONAL]
- [ ] Great job written test to asset all the cases requested in the project requirement :clap: However, some the tests in the `spec/requests/` directory are failing. Kindly make sure all the written tests in your application are passing.
Screenshot from 2023-12-19 21-44-00

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by richie1988** [comment]
## Comfirmed all tests are working and my tests implementation all working just fine has requested by the reviewer

ShareX_sBqsL53com

ShareX_Nx1xUIuWXe
Code_a8ZR8dwUab

**Review by Strangeal — APPROVED** [review]
Hi @richie1988, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: Congratulations! 🎉 



## Highlights
- Passing tests ✅

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**